site/app/libraries/database/DatabaseQueries.php

```diff
@@ -2652,4 +2674,418 @@ public function getTeamsById(array $team_ids) {
+                $this->course_db->convertBoolean($gradeable->isTaGrading()),
+                $this->course_db->convertBoolean($gradeable->isStudentView()),
+                $this->course_db->convertBoolean($gradeable->isStudentSubmit()),
+                $this->course_db->convertBoolean($gradeable->isStudentDownload()),
+                $this->course_db->convertBoolean($gradeable->isStudentDownloadAnyVersion()),
+                $gradeable->getAutogradingConfigPath(),
+                $gradeable->getLateDays(),
+                $gradeable->getPrecision(),
+                $this->course_db->convertBoolean($gradeable->isPeerGrading()),
+                $gradeable->getPeerGradeSet()
+            ];
+            $this->course_db->query("
+                INSERT INTO electronic_gradeable(
+                  g_id,
+                  eg_submission_open_date,
+                  eg_submission_due_date,
+                  eg_is_repository,
+                  eg_subdirectory,
+                  eg_team_assignment,
+                  eg_max_team_size,
+                  eg_team_lock_date,
+                  eg_use_ta_grading,
+                  eg_student_view,
+                  eg_student_submit,
+                  eg_student_download,
+                  eg_student_any_version,
+                  eg_config_path,
+                  eg_late_days,
+                  eg_precision,
+                  eg_peer_grading,
+                  eg_peer_grade_set)
+                VALUES(?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)", $params);
+        }
+
+        // Also make sure to create components
+        $this->updateGradeableComponents($gradeable);
+    }
+
+    /**
+     * Iterates through each component in a gradeable and updates/creates
+     *  it in the database as necessary.  It also reloads the marks/components
+     *  if any were added
+     * @param \app\models\gradeable\Gradeable $gradeable
+     */
+    private function updateGradeableComponents(\app\models\gradeable\Gradeable $gradeable) {
+
+        // iterate through components and see if any need updating/creating
+        $invalidateComponents = false;
+        $order = 0;
+        foreach ($gradeable->getComponents() as $component) {
+            // Rectify component order
+            if ($component->getOrder() !== $order) {
+                $component->setOrder($order);
+            }
+            ++$order;
+
+            // New component, so add it
+            if ($component->getId() < 1) {
+                $this->createComponent($component);
+                $invalidateComponents = true;
+            } else {
+                $this->updateComponent($component);
+            }
+        }
+
+        // If any components were new, recreate the components array
+        $components = $gradeable->getComponents();
+        if ($invalidateComponents) {
+            $new_components = $this->getGradeableComponentConfigs($gradeable);
```

Comment:
Any reason you have to re-query the database here? I'd recommend making `createComponent()` return the id of the added database entry so you don't have to reload to get updated ids.

Reply:
The idea was that components and marks can't have their id's changed during their lifecycle, but I definitely understand that this suboptimal.  Changes to the number of components in a gradeable is rare since only the instructor can do it and they typically only do it a few times.  Adding/deleting marks is more common and results in a similar reload.

Reply:
Yes, this won't happen often, but it still seems unnecessary to have to delete and reload all mark and component data, especially if other classes already have references to the old objects.

Reply:
Yes, this won't happen often, but it still seems unnecessary to have to delete and reload all mark and component data, especially if other classes already have references to the old objects.